src/emmy/bigfraction.cljs

```diff
@@ -0,0 +1,319 @@
+#_"SPDX-License-Identifier: GPL-3.0"
+
+(ns emmy.bigfraction
+  "A CLJS bigfraction is a coprime pair of JavaScript BigInts, with the
+   sign carried in the numerator."
+  (:refer-clojure :exclude [abs zero? neg?])
+  (:require
+   [clojure.core :as core]
+   [emmy.value :as v]
+   [goog.array :as garray]))
+
+(def ^:private ZERO (js/BigInt 0))
+(def ^:private ONE (js/BigInt 1))
+(def ^:private TEN (js/BigInt 10))
+(def ^:private -ONE (- ONE))
+
+(defn- bigint?
+  "Returns true if x is a BigInt. There is a similar function in [[emmy.util]],
+   but we prefer that this library avoid that dependency."
+  [x]
+  (= "bigint" (goog/typeOf x)))
+
+(declare eq cmp ->real)
+
+(deftype Fraction [^js/BigInt n ^js/BigInt d]
+
+  Object
+  (valueOf [this] (->real this))
+  (toString [_] (str n "/" d))
+
+  ;; these are the remaining dependencies on value.
+  ;; Maybe they should be part of extend-type in the ratio library.
+  v/IKind
+  (kind [_] :emmy.ratio/ratio)
+
+  v/INumericTower
+
+  v/IReal
+
+  v/Numerical
+  (numerical? [_] true)
+
+  IHash
+  (-hash [x]
+    (bit-xor
+     (-hash (.-n x))
+     (-hash (.-d x))))
+
+  IComparable
+  (-compare [this other]
+    (cond (instance? Fraction other)
+          (cmp this other)
+
+          (bigint? other)
+          (compare n (* d other))
+
+          :else
+          (let [o-value (.valueOf other)]
+            (if (v/real? o-value)
+              (garray/defaultCompare this o-value)
+              (throw (js/Error (str "Cannot compare " this " to " o-value " : " (v/kind o-value) " (" (type o-value) ")")))))))
+
+  IEquiv
+  (-equiv [this other]
+    (cond (instance? Fraction other) (eq this other)
+          (v/integral? other) (and (== ONE d) (v/= n other))
+          :else false))
+
+  IPrintWithWriter
+  (-pr-writer [_ writer _] (write-all writer "#emmy/ratio \"" n "/" d "\"")))
+
+(def ^:private F_ONE (Fraction. ONE ONE))
+
+(defn division-by-zero
+  "Throws JS exception used to signal an attempt to construct a fraction
+   with zero denominator."
+  []
+  (throw (js/Error "Fraction with zero denominator")))
+
+(defn numerator [^Fraction x]
+  (.-n x))
+
+(defn denominator [^Fraction x]
+  (.-d x))
+
+(defn zero?
+  "Returns true iff `x` is a zero fraction."
+  [^Fraction x]
+  (let [a (.-n x)]
+    (== ZERO a)))
+
+(defn one?
+  "Returns true iff `x` is a unit fraction."
+  [^Fraction x]
+  (let [a (.-n x)
+        b (.-d x)]
+    (== a b)))
+
+(defn eq [^Fraction x ^Fraction y]
+  (let [a (.-n x)
+        b (.-d x)
+        c (.-n y)
+        d (.-d y)]
+    (and (== a c) (== b d))))
+
+(defn bigint-gcd
+  "GCD assuming a and b are BigInts > 0"
+  [^js/BigInt a ^js/BigInt b]
+  (loop [a a
+         b b]
+    (if (== b ZERO) a
+        (recur b (core/mod a b)))))
```

Comment:
nice and core/mod works for bigint?

I think `js-mod` would be cheaper here and identical in behavior for positive numbers...

Reply:
agree, this is an oversight/will fix. Initially I didn't do the work as cljs-only, but that turned out to be a liberating way to approach it instead of hoping that core/ did what I wanted in these cases